File: ast/expression/ConstructorCallExpression.ts
import { Type } from "../../type_inference/Type";
import { SourceLocation } from "../ASTElement";
import { ExpressionElement } from "../ExpressionElement";
import { TypeElement } from "../TypeElement";


export class ConstructorCallExpression extends ExpressionElement {
    source: TypeElement;
    args: ExpressionElement[];

    constructor(loc: SourceLocation, source: TypeElement, args: ExpressionElement[]) {
        super(loc);
        this.source = source;
        this.args = [...args];
    }

    toString() {
        return `new ${this.source}(${this.args.map(x => x.toString()).join(", ")})`;
    }

    clone() {
        return new ConstructorCallExpression(
            this.source_location,
            this.source.clone(),
            this.args.map(x => x.clone()) as ExpressionElement[]
        );
    }
}

File: backend/src/main/java/dev/m0rg/howl/ast/ConstructorCallExpression.java
package dev.m0rg.howl.ast;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import dev.m0rg.howl.llvm.LLVMBuilder;
import dev.m0rg.howl.llvm.LLVMFunction;
import dev.m0rg.howl.llvm.LLVMValue;

public class ConstructorCallExpression extends CallExpressionBase {
    TypeElement source;

    public ConstructorCallExpression(Span span) {
        super(span);
    }

    @Override
    public ASTElement detach() {
        ConstructorCallExpression rc = new ConstructorCallExpression(span);
        rc.setSource((TypeElement) source.detach());
        this.copyArguments(rc);
        return rc;
    }

    @Override
    public String format() {
        return "new " + this.source.format() + this.getArgString();
    }

    public void setSource(TypeElement source) {
        this.source = (TypeElement) source.setParent(this);
    }

    public void transform(ASTTransformer t) {
        source.transform(t);
        this.setSource(t.transform(source));
        this.transformArguments(t);
    }

    @Override
    public TypeElement getType() {
        return source;
    }

    @Override
    public Map<String, FieldHandle> getUpstreamFields() {
        HashMap<String, FieldHandle> rc = new HashMap<>();
        addFields(rc);
        return rc;
    }

    @Override
    protected TypeElement getTypeForArgument(int index) {
        ClassType source_type = (ClassType) source.resolve();
        Optional<Function> constructor = source_type.getSource().getConstructor();
        if (constructor.isPresent()) {
            return constructor.get().getArgumentList().get(index + 1).getOwnType();
        }
        return new NamedType(this.span, "__error");
    }

    @Override
    public LLVMValue generate(LLVMBuilder builder) {
        ClassType source_type = (ClassType) source.resolve();
        LLVMFunction callee = source_type.getSource().allocator;
        List<LLVMValue> args = new ArrayList<>(this.args.size());
        for (Expression e : this.args) {
            args.add(e.generate(builder));
        }
        return builder.buildCall(callee, args, "");
    }
}
